[refactor] simplify Source Code based on DOM API & TypeScript

TechQuery opened the pull request:
## Major changes

1. [refactor] simplify Source Code based on DOM API & TypeScript
2. [refactor] simplify Test cases of Markdown converting
3. [add] an optional parameter for Custom Ignore elements
4. [optimize] simplify Input Markdown format
5. [optimize] output Markdown with YAML front-matter
6. [optimize] add an Issue Comment after fetching done
7. [optimize] upgrade Upstream packages & update Read Me documents
8. [fix] "srcset" attribute converting (mixmark-io/turndown#305)
9. [migrate] replace NPM with PNPM to reduce Disk Usage

## Test repository

https://github.com/TechQuery/Web-fetch-action-test

TechQuery committed:
[migrate] replace NPM with PNPM to reduce Disk Usage
[optimize] upgrade Upstream packages

TechQuery committed:
[refactor] simplify Source Code based on JSDOM

stevending1st commented on package.json:
```diff
@@ -1,13 +1,14 @@
     "jest": "jest",
     "prefix": "lint-staged",
-    "build": "ncc build ./src/index.js -o dist --license licenses.txt"
+    "build": "rm -rf dist/  &&  ncc build ./src/index.js -m -o dist --license licenses.txt"
```
安装 `rimraf` 再使用其删除

stevending1st commented on src/index.js:
```diff
@@ -1,52 +1,46 @@
+        ) || {};
+
+    const articleText = `> -  原文地址：[${title}](${path})
+> -  原文作者：[${textContent.trim() || '匿名'}](${href})
```
没有作者理论上可以不显示吧。

luojiyin1987 replied:
除了官方的编辑人员工作出错。 见过没名的。去 fcc  的官方 GitHub 项目 提 issue，让官方的人去复核。

TechQuery committed:
[migrate] refactor HTML parser based on TypeScript & DOM API

luojiyin1987 commented on package.json:
```diff
@@ -1,13 +1,14 @@
+  "main": "dist/index.js",
   "scripts": {
+    "prepare": "husky install",
     "jest": "jest",
     "prefix": "lint-staged",
-    "build": "ncc build ./src/index.js -o dist --license licenses.txt"
+    "build": "rimraf dist/  &&  ncc build ./src/index.ts -m -o dist --license licenses.txt"
```
如果  npm 安装不成功，或者 node_modules 文件夹损坏，rimraf 缺少对应的依赖，就运行不成功。

TechQuery replied:
> 如果  npm 安装不成功，或者 node_modules 文件夹损坏，rimraf 缺少对应的依赖，就运行不成功。 

但为什么会有这些情况出现呢？

luojiyin1987 replied:
在 CI 环境中，很难出现。即使出现，也可以根据日志定位出。

TechQuery replied:
> 在 CI 环境中，很难出现。即使出现，也可以根据日志定位出。

所以我没明白这里有什么问题？

TechQuery committed:
[migrate] replace JSDOM with LinkeDOM to reduce Bundle Size

TechQuery committed:
[fix] "srcset" attribute converting (mixmark-io/turndown#305)

TechQuery committed:
[refactor] simplify Test cases of Markdown converting
[optimize] Picture Source detecting of Markdown converting
[optimize] update Read Me documents

TechQuery committed:
[add] an optional parameter for Custom Ignore elements
[optimize] simplify Input Markdown format
[optimize] output Markdown with YAML front-matter
[optimize] add an Issue Comment after fetching done

TechQuery committed:
[optimize] update Read Me document

TechQuery commented:
@stevending1st @luojiyin1987 review 并合并后，再创建一个 `v1` 分支（并设为默认分支），方便他人引用。

luojiyin1987 reviewed (commented):
为什么从 cheerio  换到 JSDOM ，再到 linkedom ？

TechQuery commented:
> 为什么从 cheerio  换到 JSDOM ，再到 linkedom ？

因为 cheerio 早已不再维护，其复刻的 jQuery API 的核心内容也早已融入了标准 DOM API，不需要再做这一层封装，也就是 JSDOM 所实现的。但又因为它的实现过于庞杂，使得体积过大，就又有人写了超轻量级的 LinkeDOM。
